### PR TITLE
Adding temperature decoding option

### PR DESCRIPTION
**Pull request by spencerp**
**Patch description**
Basing this change on the definition of temperature in https://arxiv.org/pdf/1503.02531.pdf and the implementation found in fairseq: https://github.com/pytorch/fairseq/blob/master/fairseq/sequence_generator.py#L573-L574

**Testing steps**
Add breakpoint and verify temperature is applied to the pre-softmax logits.
```
λ python parlai/scripts/self_chat.py -mf /path/to/model   --skip-generation False --beam-block-ngram 1 --beam-context-block-ngram 1 --inference topk --topk 4 --selfchat-max-turns 6 --num-self-chats 1 -d True --beam-size 2 --temperature 100
...
[creating task(s): self_chat]
   [context]: __SILENCE__
> /private/home/spoff/ParlAI/parlai/core/torch_generator_agent.py(896)_generate()
-> if self.temperature != 1.0:
(Pdb) F.log_softmax(score, dim=-1)
tensor([[[-1.6266e+01, -6.5504e+04, -8.9844e+00,  ..., -1.6250e+01,
          -1.6266e+01, -1.6250e+01],
         [-1.6266e+01, -6.5504e+04, -8.9844e+00,  ..., -1.6250e+01,
          -1.6266e+01, -1.6250e+01]]], device='cuda:0', dtype=torch.float16)
(Pdb) F.log_softmax(score.div(self.temperature), dim=-1)
tensor([[[ -10.9297, -666.0000,  -10.8594,  ...,  -10.9297,  -10.9297,
           -10.9297],
         [ -10.9297, -666.0000,  -10.8594,  ...,  -10.9297,  -10.9297,
           -10.9297]]], device='cuda:0', dtype=torch.float16)
(Pdb) n
> /private/home/spoff/ParlAI/parlai/core/torch_generator_agent.py(897)_generate()
-> score.div_(self.temperature)
(Pdb) n
> /private/home/spoff/ParlAI/parlai/core/torch_generator_agent.py(898)_generate()
-> score = F.log_softmax(score, dim=-1)
(Pdb) n
> /private/home/spoff/ParlAI/parlai/core/torch_generator_agent.py(899)_generate()
-> for i, b in enumerate(beams):
(Pdb) score
tensor([[[ -10.9297, -666.0000,  -10.8594,  ...,  -10.9297,  -10.9297,
           -10.9297],
         [ -10.9297, -666.0000,  -10.8594,  ...,  -10.9297,  -10.9297,
           -10.9297]]], device='cuda:0', dtype=torch.float16)
```

### CODE REVIEW

File: parlai/core/torch_generator_agent.py [hunk 1]
```diff
@@ -352,6 +358,8 @@ def __init__(self, opt: Opt, shared=None):
         self.beam_min_length = opt.get('beam_min_length', 1)
         self.beam_block_ngram = opt.get('beam_block_ngram', -1)
         self.beam_context_block_ngram = opt.get('beam_context_block_ngram', -1)
+        self.temperature = opt.get('temperature')
```
Comment (stephenroller) [inline parlai/core/torch_generator_agent.py]: set a default of 1 so mypy doesn't complain

File: parlai/core/torch_generator_agent.py [hunk 2]
```diff
@@ -1123,7 +1133,10 @@ def advance(self, logprobs):
         #  check new hypos for eos label, if we have some, add to finished
         for hypid in range(self.beam_size):
             if self.outputs[-1][hypid] == self.eos:
-                if self.scores[hypid] == neginf(self.scores.dtype):
+                if (
+                    self.scores[hypid] == neginf(self.scores.dtype)
+                    or self.scores[hypid].item() == -math.inf
```
Comment (stephenroller) [inline parlai/core/torch_generator_agent.py]: this slipped in. i assume you want it as part of another PR still?
